src/pydrex/diagnostics.py

```diff
@@ -0,0 +1,193 @@
+    # SciPy returns eigenvalues in ascending order (same order for vectors).
+    a_mean = np.zeros((3, 3))
+    a_mean[0, 0] = np.sum(orientations[:, 0, 0] ** 2)
+    a_mean[1, 1] = np.sum(orientations[:, 0, 1] ** 2)
+    a_mean[2, 2] = np.sum(orientations[:, 0, 2] ** 2)
+    a_mean[0, 1] = np.sum(orientations[:, 0, 0] * orientations[:, 0, 1])
+    a_mean[0, 2] = np.sum(orientations[:, 0, 0] * orientations[:, 0, 2])
+    a_mean[1, 2] = np.sum(orientations[:, 0, 1] * orientations[:, 0, 2])
+    _, a_eigenvectors = la.eigh(_tensors.upper_tri_to_symmetric(a_mean))
+    a_vector = a_eigenvectors[:, -1]
+
+    b_mean = np.zeros((3, 3))
+    b_mean[0, 0] = np.sum(orientations[:, 1, 0] ** 2)
+    b_mean[1, 1] = np.sum(orientations[:, 1, 1] ** 2)
+    b_mean[1, 2] = np.sum(orientations[:, 1, 2] ** 2)
+    b_mean[0, 1] = np.sum(orientations[:, 1, 0] * orientations[:, 1, 1])
+    b_mean[0, 2] = np.sum(orientations[:, 1, 0] * orientations[:, 1, 2])
+    b_mean[1, 2] = np.sum(orientations[:, 1, 1] * orientations[:, 1, 2])
+    _, b_eigenvectors = la.eigh(_tensors.upper_tri_to_symmetric(b_mean))
+    b_vector = b_eigenvectors[:, -1]
+
+    c_mean = np.zeros((3, 3))
+    c_mean[0, 0] = np.sum(orientations[:, 2, 0] ** 2)
+    c_mean[1, 1] = np.sum(orientations[:, 2, 1] ** 2)
+    c_mean[2, 2] = np.sum(orientations[:, 2, 2] ** 2)
+    c_mean[0, 1] = np.sum(orientations[:, 2, 0] * orientations[:, 2, 1])
+    c_mean[0, 2] = np.sum(orientations[:, 2, 0] * orientations[:, 2, 2])
+    c_mean[1, 2] = np.sum(orientations[:, 2, 1] * orientations[:, 2, 2])
+    _, c_eigenvectors = la.eigh(_tensors.upper_tri_to_symmetric(c_mean))
+    c_vector = c_eigenvectors[:, -1]
+
+    return (
+        a_vector / la.norm(a_vector),
+        b_vector / la.norm(b_vector),
+        c_vector / la.norm(c_vector),
+    )
+
+
+def resample_orientations(orientations, fractions, n_samples=None):
+    """Generate samples from `orientations` based on volume distribution.
+
+    If the optional number of samples is not specified,
+    it will be set to the number of original "grains" (length of `fractions`).
+
+    """
+    if n_samples is None:
+        n_samples = len(fractions)
+    sort_ascending = np.argsort(fractions)[::-1]
+    frac_ascending = np.take(fractions, sort_ascending)
+    # Cumulative volume fractions.
+    cumfrac = np.cumsum(frac_ascending)
+    # Number of new samples with volume less than each cumulative fraction.
+    count_less = [np.searchsorted(cumfrac, r) for r in rn.random_sample(n_samples)]
+    return np.stack([orientations[i] for i in count_less])
+
+
+def M_index(orientations):
+    """Calculate M-index for olivine polycrystal orientations.
+
+    See [Skemer et al. 2005].
+
+    [Skemer et al. 2005]: https://doi.org/10.1016/j.tecto.2005.08.023
+
+    """
+    misorientations = np.fromiter(
+        (misorientation_angle(A, B) for A, B in it.combinations(orientations, 2)),
+        dtype=float,
+    )
```

Comment:
Any reason not to use
`misorientations = [misorientation_angle(A, B) for A, B in it.combinations(orientations, 2)]`
?